primitives/src/settlement.rs

```diff
@@ -0,0 +1,437 @@
+pub struct Venue {
+    /// Identity of the venue's creator
+    pub creator: IdentityId,
+    /// Specifies type of the venue (Only needed for the UI)
+    pub venue_type: VenueType,
+}
+
+/// Details about an offchain transaction receipt
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug, PartialOrd, Ord)]
+pub struct Receipt<Balance> {
+    /// Unique receipt number set by the signer for their receipts
+    pub receipt_uid: u64,
+    /// Identity of the sender
+    pub from: PortfolioId,
+    /// Identity of the receiver
+    pub to: PortfolioId,
+    /// Ticker of the asset being transferred
+    pub asset: Ticker,
+    /// Amount being transferred
+    pub amount: Balance,
+}
+
+/// A wrapper for VenueDetails
+#[derive(Encode, Decode, TypeInfo, VecU8StrongTyped)]
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ReceiptMetadata(Vec<u8>);
+
+/// Details about an offchain transaction receipt that a user must input
+#[derive(Encode, Decode, TypeInfo, Clone, PartialEq, Eq, Debug, PartialOrd, Ord)]
+pub struct ReceiptDetails<AccountId, OffChainSignature> {
+    /// Unique receipt number set by the signer for their receipts
+    pub receipt_uid: u64,
+    /// Target leg id
+    pub leg_id: LegId,
+    /// Signer for this receipt
+    pub signer: AccountId,
+    /// signature confirming the receipt details
+    pub signature: OffChainSignature,
+    /// Generic text that can be used to attach messages to receipts
+    pub metadata: ReceiptMetadata,
+}
+
+/// Stores the number of fungible and non fungible transfers in a set of legs.
+#[derive(Default)]
+pub struct TransferData {
+    fungible: u32,
+    non_fungible: u32,
+}
+
+impl TransferData {
+    /// Creates an instance of `TransfersData`.
+    pub fn new(fungible: u32, non_fungible: u32) -> Self {
+        TransferData {
+            fungible,
+            non_fungible,
+        }
+    }
+
+    /// Returns the number of fungible transfers.
+    pub fn fungible(&self) -> u32 {
+        self.fungible
+    }
+
+    /// Returns the number of non fungible transfers.
+    pub fn non_fungible(&self) -> u32 {
+        self.non_fungible
+    }
+
+    /// Adds one to the number of fungible transfers.
+    pub fn add_fungible(&mut self) {
+        self.fungible += 1;
+    }
+
+    /// Adds `nfts.len()` to the number of non fungible transfers.
+    pub fn add_non_fungible(&mut self, nfts: &NFTs) {
+        self.non_fungible += nfts.len() as u32;
+    }
+
+    /// Gets the `TransferData` from a slice of `LegV2`.
+    pub fn from_legs(legs_v2: &[LegV2]) -> Result<TransferData, String> {
+        let mut transfer_data = TransferData::default();
+        for leg_v2 in legs_v2 {
+            match &leg_v2.asset {
+                LegAsset::Fungible { .. } => {
+                    if transfer_data.fungible().checked_add(1).is_none() {
+                        return Err(String::from(
+                            "Number of fungible assets is greater than allowed",
```

Comment:
I think we should use a much lower limit then `u32::MAX`.  There is no way to encode 4 Billion legs in a single transaction.  We can still use a value higher (1024?) than the chain runtime limit for number of legs and number of fungible assets.

Also might be a little nicer to put the `checked_add` in the `add_fungible` method.

Reply:
Yeah, I considered that. I wanted to avoid having to deal with that in the `get_transfer_data` which is used in the extrinsics and will never get close to the limits